### DISCUSSION TITLE
A gold mine in waiting

### DISCUSSION THREAD

**Discussion by NewsSoup** [pyscript/pyscript#1600] [Proposals]
I am putting this under feature proposals, but I am actually proposing an offshoot of PyScript that can serve to expand your product line. I offer it for your consideration, as it is way above my skill level.

I have recently gotten some heavy experience in creating both a website and a desktop app for a research assistance service. Ever since, I have often wondered why nobody has ever designed a window-based GUI system that acts like a browser in rendering window content from HTML, CSS and JavaScript. I figured the issue was integration of HTML and Python (or another programming language) - a problem that you appear to have already solved...

So my proposal is this: in order to expand you product line in a way where both products integrate better with web design as they grow, the PyScript team should consider creating a browser-like window-based GUI system that uses HTML, CSS and JavaScript for basic window layout and style and creates python objects of the GUI elements that can them be manipulated in Python.

This could create a powerful system for businesses to easily expand their capabilities without the need to retrain web developers to develop desktop GUIs. It would certainly earn you name recognition, and as long as rendering is based on browser industry standards it gives you a synergistic platform to further perfect PyScript. The Python objects need to be able to read and write to the HTML, so the implementation is not actually difficult.

**Comment by WebReflection** [comment]
is this some sort of React native like approach? With Pythonic UI elements (API) that can trnaslate into Gtk and/or Web? Not sure I am following, mostly because to me having access to the DOM already gives you that freedom to manipulate "HTML" elements.

**Reply by NewsSoup** [reply]
I'm approaching it from a desktop app perspective. What we have with web development is really convenient, the system of HTML, CSS and JavaScript is really powerful for creating GUIs with ease. However, it exists only for web browsers, which is unfortunate. There is no reason that a window can't render it like a browser, but without the web connections and certificates.

So no, not Python GUI elements, but rather a reduced browser that doesn't connect to the internet but only renders the GUI. The backend interacts with the GUI using normal python code, as if it were Pythonic GUI elements. To merge the two, you need to create objects with the browser code that uses PyScript to manage the DOM on the browser, but acts like a Python object.

In this way, PyScript can be used for both web and desktop development, and this versatility might help you reach a broader market. I'm genuinely talking of just a browser window to render a GUI, it can't be too hard given that only the rendering elements are needed.

**Reply by WebReflection** [reply]
the Web *is* hard though, so you end up needing a WebView to simplify what you are after (if I understood correctly) ... but that looks like a Tauri based app (which is something like Electron except it has only the view) that communicates with a Python backend able to reach the UI ... which is what [coincident/server](https://github.com/WebReflection/coincident#coincidentserver) offers and *coincident* module is what fuels PyScript Next (core) so that theoretically we should be able to orchestrate something similar with Python only but we're not there yet.

**Comment by Neon22** [comment]
Do you mean like pypercard ?
 - https://pypercard.readthedocs.io/en/latest/tutorial1.html

**Reply by NewsSoup** [reply]
That looks cool, I will look into it. But from a preliminary view it doesn't give you the ability to style with CSS, which would make it very easy to do complex window layouts with very little rendering time.

**Reply by WebReflection** [reply]
Apologies I've thought this was a reply to my latest comment ... please ignore my previous comments, thanks.
